build-system/tasks/helpers.js

```diff
@@ -374,96 +371,119 @@ function finishBundle(srcFilename, destDir, destFilename, options) {
-            internalRuntimeVersion,
-            'runtime-version'
-          )
-        )
-        .pipe(regexpSourcemaps(/([^]+)/, devWrapper, 'wrapper'))
-        .pipe(rename(destFilename))
-        .pipe(sourcemaps.write('.'))
-        .pipe(gulp.dest(destDir))
-        .on('end', () =>
-          finishBundle(srcFilename, destDir, destFilename, options)
-        )
-    )
-      .then(() => {
-        let name = destFilename;
-        if (options.latestName) {
-          const latestMaxName = options.latestName.split('.js')[0] + '.max.js';
-          name = `${name} → ${latestMaxName}`;
-        }
-        endBuildStep('Compiled', name, startTime);
-      })
-      .then(() => {
-        const target = path.basename(destFilename, path.extname(destFilename));
-        if (UNMINIFIED_TARGETS.includes(target)) {
-          return applyAmpConfig(
-            `${destDir}/${destFilename}`,
-            /* localDev */ true,
-            /* fortesting */ options.fortesting
-          );
+  const esbuildBabelPlugin = {
+    name: 'babel',
+    async setup(build, {transform} = {}) {
+      const updateVersion = (contents) => {
+        return contents.replace(
+          /\$internalRuntimeVersion\$/g,
+          internalRuntimeVersion
+        );
+      };
+
+      const transformContents = async ({file, contents}) => {
+        const babelOptions = babel.loadOptions({
+          caller: {name: 'unminified'},
+          filename: file.path,
+          sourceFileName: path.relative(process.cwd(), file.path),
+        });
+        const result = await babel.transformAsync(contents, babelOptions);
+        return {contents: updateVersion(result.code)};
+      };
+
+      if (transform) {
+        return await transformContents(transform);
+      }
+
+      build.onLoad({filter: /.*/, namespace: ''}, async (file) => {
+        if (cache.has(file.path)) {
+          return {contents: cache.get(file.path)};
         }
+        const contents = await fs.promises.readFile(file.path, 'utf-8');
+        const transformed = await transformContents({file, contents});
+        cache.set(file.path, transformed.contents);
+        return transformed;
       });
+    },
+  };
+
+  /**
+   * Splits up the wrapper to compute the banner and footer
+   * @return {Object}
+   */
+  function processWrapper() {
+    const wrapper = options.wrapper || wrappers.none;
+    const sentinel = '<%= contents %>';
+    const start = wrapper.indexOf(sentinel);
+    return {
+      banner: wrapper.slice(0, start),
+      footer: wrapper.slice(start + sentinel.length),
+    };
+  }
+
+  /**
+   * Bundles an entry point with all its imports and applies babel transforms.
+   * @param {boolean} continueOnError
+   */
+  async function performBundle(continueOnError) {
+    const {banner, footer} = processWrapper();
+    await esbuild
+      .build({
+        entryPoints: [entryPoint],
+        bundle: true,
+        sourcemap: true,
+        define: experimentDefines,
```

Comment:
Can we add `$internalRuntimeVersion$` to the define set?

Reply:
Tried doing this in a few different ways, none seemed to work. I think the double quotes around `$internalRuntimeVersion$` are messing with `esbuild`. Escaping the double quotes didn't work either. Will punt on this for now and see if there are other tricks to use during the `gulp dist` rewrite.

Reply:
I don't believe you need the quotation marks

Reply:
Sorry, I edited my previous comment with more info. Doing this without quotes results in no substitution.

The original code is:

https://github.com/ampproject/amphtml/blob/ee68dd7be3707618d7c077d493e9ad3778a97a84/src/internal-version.js#L26-L28

And `esbuild` converts that code to:

![image](https://user-images.githubusercontent.com/26553114/108111902-5ed4a700-7063-11eb-8de3-603cef237e8c.png)

Perhaps `define` only works for constant identifiers, and not string literals?

Reply:
Darn. I think you're right. Can punt to a future cleanup PR but do you think it'd be reasonable to convert `internalRuntimeVersion` into a constant?

Reply:
Good idea, it might be possible to make it a const, but I think that's best left for a follow up PR where the change can be tested in isolation. Will do soon.

Reply:
+1 for a followup, also worth checking if this is an outstanding feature request to `esbuild`.

Reply:
Changing this to a const is more complicated than initially assumed due to https://github.com/ampproject/amphtml/pull/32663#discussion_r577210915. Moving it to a babel plugin seems to be the best option.